### PR TITLE
Pt-633: Enrolment can't start after event start and not in the past

### PR DESCRIPTION
**Pull request by nenonja**
## Description :sparkles:
Currently there is no validation for enrolment date. These are added:
- enrolment cannot start in the past
- enrolment cannot start after the event has started
Both validation checks the inputted hours too.

### Closes :no_good_woman:
https://helsinkisolutionoffice.atlassian.net/browse/PT-633

## Screenshots :camera_flash:
Previously (no errors):
Screenshot 2021-01-27 at 12 32 02 PM
After:
- Enrolment cannot start in the past
Screenshot 2021-01-27 at 12 19 46 PM
- enrolment cannot start after the event has started
Screenshot 2021-01-27 at 12 20 17 PM

### CODE REVIEW

**Comment by nenonja** [comment]
TimePicker throws following error, it might have something wrong:
```
    console.error node_modules/react-dom/cjs/react-dom.development.js:88
      Warning: A component is changing an uncontrolled input of type undefined to be controlled. Input elements should not switch from uncontrolled to controlled (or vice versa). Decide between using a controlled or uncontrolled input element for the lifetime of the component. More info: https://fb.me/react-controlled-components
          in input (at Timepicker.tsx:110)
          in div (at Timepicker.tsx:109)
          in div (at InputWrapper.tsx:78)
          in div (at InputWrapper.tsx:49)
          in ForwardRef (at Timepicker.tsx:100)
          in Timepicker (at TimepickerField.tsx:47)
          in InputField (created by Field)
          in Field (at EventForm.tsx:253)
          in div (at EventForm.tsx:247)
          in div (at FormGroup.tsx:7)
          in FormGroup (at EventForm.tsx:246)
          in div (at EventForm.tsx:236)
          in div (at EventForm.tsx:186)
          in div (at EventForm.tsx:185)
          in form (at EventForm.tsx:184)
          in Formik (at EventForm.tsx:141)
          in EventForm (at EventForm.test.tsx:41)
          in Router (at testUtils.tsx:35)
          in ApolloProvider (created by MockedProvider)
          in MockedProvider (at testUtils.tsx:34)
          in Provider (at testUtils.tsx:33)
          in Wrapper

```

File: src/domain/event/eventForm/ValidationSchema.ts [hunk 1]
```diff
@@ -35,9 +38,35 @@ const createValidationSchemaYup = (
...
+              : occurenceDate;
+            return schema.test(
+              'isBefore',
+              () => ({
+                key: VALIDATION_MESSAGE_KEYS.DATE_MAX,
+                max: formatDate(occurenceStart, DATETIME_FORMAT),
+              }),
+              (enrolmentStart: Date) => {
+                return enrolmentStart < occurenceStart;
+              }
```
Comment (Sartonon) [inline src/domain/event/eventForm/ValidationSchema.ts:65]: Very minor but could use implicit return

```
 (enrolmentStart: Date) => enrolmentStart < occurenceStart;
```

File: src/domain/occurrence/__tests__/CreateOccurrencePage.test.tsx [hunk 2]
```diff
@@ -130,10 +129,20 @@ afterEach(() => {
   jest.restoreAllMocks();
 });
 
-test('renders coming occurrences table correctly', async () => {
-  initializeMocks();
-  renderWithRoute(<CreateOccurrencePage />, {
+const InitializeMocksAndRenderPage = (
```
Comment (Sartonon) [inline src/domain/occurrence/__tests__/CreateOccurrencePage.test.tsx:132]: Is the capital letter at start intentional? :)

File: src/domain/event/eventForm/ValidationSchema.ts [hunk 3]
```diff
@@ -35,9 +38,35 @@ const createValidationSchemaYup = (
...
-        VALIDATION_MESSAGE_KEYS.DATE_TODAY_OR_LATER,
-        isTodayOrLater
+        'isInTheFuture',
+        VALIDATION_MESSAGE_KEYS.DATE_IN_THE_FUTURE,
+        isFuture
+      )
+      .when(
+        ['occurrenceDate', 'occurrenceStartsAt'],
+        (
+          occurenceDate: Date,
```
Comment (Sartonon) [inline src/domain/event/eventForm/ValidationSchema.ts:48]: Small typo with occurrurrence :) .....hah even I can't write it :D

File: src/domain/occurrence/__tests__/CreateOccurrencePage.test.tsx [hunk 4]
```diff
@@ -130,10 +129,20 @@ afterEach(() => {
...
+  renderOptions,
+  fromDate = new Date(2020, 7, 2),
+  occurences = 5
+) => {
+  initializeMocks(fromDate, occurences);
```
Comment (Sartonon) [inline src/domain/occurrence/__tests__/CreateOccurrencePage.test.tsx:137]: After this refactor there is no longer need for mocks to be initialized to global scope of this file. They could be return form this function now. I think that approach would be easier to understand.

**Review by Sartonon — COMMENTED** [review]
Good job! Couple comments added.

File: src/domain/event/eventForm/ValidationSchema.ts (hunk 1 above)
Comment (nenonja) [reply src/domain/event/eventForm/ValidationSchema.ts]: fixed, although sometimes is easier to use longer format so it's easy to throw in console.log when debugging.

File: src/domain/occurrence/__tests__/CreateOccurrencePage.test.tsx (hunk 2 above)
Comment (nenonja) [reply src/domain/occurrence/__tests__/CreateOccurrencePage.test.tsx]: fixed

File: src/domain/event/eventForm/ValidationSchema.ts (hunk 3 above)
Comment (nenonja) [reply src/domain/event/eventForm/ValidationSchema.ts]: fixed, the typo was in many places

File: src/domain/occurrence/__tests__/CreateOccurrencePage.test.tsx (hunk 4 above)
Comment (nenonja) [reply src/domain/occurrence/__tests__/CreateOccurrencePage.test.tsx]: fixed

File: src/domain/event/eventForm/ValidationSchema.ts (hunk 1 above)
Comment (Sartonon) [reply src/domain/event/eventForm/ValidationSchema.ts]: one trick is to use comma operator for console.log:

`(enrolmentStart: Date) => (console.log(enrolmentStart), enrolmentStart < occurenceStart);`

**Review by Sartonon — APPROVED** [review]
Good!

**Pull request merged**